x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/preprocessing/Multi.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ml.inference.preprocessing;
+
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import org.apache.lucene.util.RamUsageEstimator;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.core.ml.utils.NamedXContentObjectHelper;
+
+/**
+ * Multi-PreProcessor for chaining together multiple processors
+ */
+public class Multi implements LenientlyParsedPreProcessor, StrictlyParsedPreProcessor {
+
+    public static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(Multi.class);
+    public static final ParseField NAME = new ParseField("multi_encoding");
+    public static final ParseField PROCESSORS = new ParseField("processors");
+    public static final ParseField CUSTOM = new ParseField("custom");
+
+    private static final ObjectParser<Multi.Builder, PreProcessorParseContext> STRICT_PARSER = createParser(false);
+    private static final ObjectParser<Multi.Builder, PreProcessorParseContext> LENIENT_PARSER = createParser(true);
+
+    private static ObjectParser<Multi.Builder, PreProcessorParseContext> createParser(boolean lenient) {
+        ObjectParser<Multi.Builder, PreProcessorParseContext> parser = new ObjectParser<>(
+            NAME.getPreferredName(),
+            lenient,
+            Multi.Builder::new
+        );
+        parser.declareNamedObjects(Multi.Builder::setProcessors,
+            (p, c, n) -> lenient ?
+                p.namedObject(LenientlyParsedPreProcessor.class, n, PreProcessor.PreProcessorParseContext.DEFAULT) :
+                p.namedObject(StrictlyParsedPreProcessor.class, n, PreProcessor.PreProcessorParseContext.DEFAULT),
+            (multiBuilder) -> multiBuilder.setOrdered(true),
```

Comment:
I do not understand why we need to `setOrdered`. I think I also do not understand what `orderedModeCallback` is all about, it seems to me it's just a way to set some additional property on the created objects. And I find it strange that the other version of `declareNamedObjects` that doesn't take an `orderedModeCallback` passes in one that throws if the value is an array given it's `declareNamedObjects` plural. Not really this PR's issue but I would like to clear this up. Could you help us with this @nik9000 please?

Reply:
@dimitris-athanasiou a user can declare named objects in two ways:

- a hashmap keyed by the named object names (unordered)
- an array of objects (ordered)

If I did not pass this call back (and subsequently check that it was called in the builder), the user could pass an unordered hashmap, which we don't allow. 

Example of unordered :
```
{
    "multi_encoding": {
        "processors":
            {
                "ngram_encoding" : {
                    "field": "foo",
                    "feature_prefix": "f",
                    "n_grams": [1],
                    "length": 2   
                },
                "one_hot_encoding": {
                    "field": "f.11", //OUTPUT from earlier ngram_encoding
                    "hot_map": {"a": "col_a"}
                }
        }
}
```

^ This is a valid named objects collection, but obviously, its order is not guaranteed. We need to throw in this case. 


As for making the declareNamedObjects array parsing more clear, that is probably best for another PR.